lib/Sema/TypeCheckConcurrency.cpp

```diff
@@ -1785,6 +1785,74 @@ static void noteGlobalActorOnContext(DeclContext *dc, Type globalActor) {
   }
 }
 
+/// Find the original type of a value, looking through various implicit
+/// conversions.
+static Type findOriginalValueType(Expr *expr) {
+  do {
+    expr = expr->getSemanticsProvidingExpr();
+
+    if (auto inout = dyn_cast<InOutExpr>(expr)) {
+      expr = inout->getSubExpr();
+      continue;
+    }
+
+    if (auto ice = dyn_cast<ImplicitConversionExpr>(expr)) {
+      expr = ice->getSubExpr();
+      continue;
+    }
+
+    if (auto open = dyn_cast<OpenExistentialExpr>(expr)) {
+      expr = open->getSubExpr();
+      continue;
+    }
+
+    break;
+  } while (true);
+
+  return expr->getType()->getRValueType();
+}
+
+bool swift::diagnoseApplyArgSendability(ApplyExpr *apply) {
```

Comment:
I think we don't need to store the `DeclContext*` in the ApplyExpr.  Instead, take the `DeclContext*` as an argument here. Then, call `SILFunction::getDeclContext()` to retrieve the one corresponding to the instruction associated with an ApplyExpr.

Reply:
This is what I was trying to get at with my earlier comments, thanks Kavon